app/services/foreman/renderer/scope/macros/base.rb

```diff
@@ -124,8 +124,12 @@ def save_to_file(filename, content, verbatim: false)
               content = Base64.encode64(content)
               "cat << #{delimiter} | base64 -d > #{filename}\n#{content}#{delimiter}"
             else
-              content += "\n" unless content.end_with?("\n")
-              "cat << #{delimiter} > #{filename}\n#{content}#{delimiter}"
+              content_echos = content.split("\n").map do |content_line|
+                "echo #{Shellwords.shellescape(content_line)} >> #{filename}"
```

Comment:
Previously the code allowed to use variables like `$var` but now that's going to be escaped, making it backwards incompatible. If you want it really verbatim, we have the base64 method above.

Reply:
This should be possible now

Reply:
Now the `verbatim` will escape everything and the `else` will escape only double quotes.

Reply:
@adamruzicka you initially introduced this because you found a case where it broke. Do you think this is sufficient?

Reply:
I don't really see where this wouldn't work, but I don't have a good feeling about it.

As a counter-proposal, how about we keep base64, but on the ruby side we drop the newlines from it and then `echo` it? Something like 
```ruby
content = Base64.encode64(content).tr("\n", "")
"echo #{content} | base64 -d > #{filename}
```

Reply:
The line can get quite long. Is there any limit that we may hit with too long lines?

Reply:
There doesn't seem to be, although shell might need a lot of time and memory to parse the script once we get into the hundreds of megabytes range. As far as I know, all the inputs passed into save_by_file are fairly small, right?

Reply:
@adamruzicka https://serverfault.com/questions/163371/linux-command-line-character-limit says that there is some limit.

I agree that it's quite long and we probably won't hit it soon, but I preferred not to rely on any of it.
I have tried to work with base64, but somehow I wasn't able to make it work properly with multiple lines and variable name. I can use a tempfile, but it seems a bit of overkill.

Reply:
> says that there is some limit.

Well, yes and no. From what I've seen, what's written there applies when launching things that are external to the shell, but it doesn't seem to apply to shell builtins, such as echo in this case.

> I have tried to work with base64, but somehow I wasn't able to make it work properly with multiple lines and variable name

How about we mix the two approaches?

```
b64 = Base64.encode64(content)
content_echos = b64.each_line.map { |l| "echo #{l}" }.join
"(#{content_echos}) | base64 -d >#{filename}"
```

Reply:
If you go the `echo` path, you need to accumulate the result somewhere. and once you have the whole base64 string accumulated, you can pass it into `base64`. 
Let me try to accumulate the base64 value without newlines and then pipe it into base64.
Something like:
```
b64 = Base64.encode64(content)
content_var = b64.each_line.map { |l| "FOO+=#{l}" }.join("\n")
content_var + "echo $FOO | base64 -d >#{filename}"
```
Mostly I had trouble with the line endings. Maybe in this approach it will be a bit easier.

But let's say this works, we will have different implementations for `verbatim` and `else` use cases.
On the other hand, if we go with the escaping approach, we will be able to DRY the code by extracting most of the generation into a single method, and just changing the escaping method of the input.

Reply:
> If you go the echo path, you need to accumulate the result somewhere

Well, you can always spawn a subshell just like I did in the previous suggestion.

> On the other hand, if we go with the escaping approach, we will be able to DRY the code by extracting most of the generation into a single method, and just changing the escaping method of the input.

True, at the same time we lose the option to write out binary blobs

Reply:
Reimplemented as base64 single string.